crates/lang/src/parse.rs

```diff
@@ -549,13 +549,180 @@ fn seek_to<'a, R: pest::RuleType>(
+            database_name: "default".to_string(),
+            table_name: "".to_string(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TablePlaceKind {
+    Local,
+    Remote(RemoteDbInfo),
+}
+
+#[derive(Debug)]
+pub struct TablePlaceKindContext {
+    pub place_kind: TablePlaceKind,
+}
+
+impl TablePlaceKindContext {
+    fn mut_remote_info<'a>(&'a mut self) -> Option<&'a mut RemoteDbInfo> {
+        if let TablePlaceKind::Remote(place_kind) = &mut self.place_kind {
+            Some(place_kind)
+        } else {
+            None
+        }
+    }
+}
+
+pub fn parse_table_place(pair: Pair<Rule>) -> LangResult<TablePlaceKindContext> {
+    let mut ctx = TablePlaceKindContext {
+        place_kind: TablePlaceKind::Local,
+    };
+    ctx.parse(pair)?;
+    Ok(ctx)
+}
+
+#[inline]
+fn parse_ip(pair: Pair<Rule>) -> LangResult<Option<IpAddr>> {
+    for p in pair.into_inner() {
+        match p.as_rule() {
+            Rule::ipv4_lit | Rule::ipv6_lit => {
+                let ip: IpAddr = p
+                    .as_str()
+                    .parse()
+                    .map_err(|e| LangError::WrappingParseIpAddrError(e))?;
+                return Ok(Some(ip));
+            }
+            _ => {}
+        }
+    }
+    Ok(None)
+}
+
+#[inline]
+fn parse_ip_address(pair: Pair<Rule>) -> LangResult<RemoteAddr> {
+    let mut addr = RemoteAddr::default();
+
+    for p in pair.into_inner() {
+        match p.as_rule() {
+            Rule::ip => {
+                addr.ip_addr = parse_ip(p)?;
+            }
+            Rule::port => {
+                let port: u16 = p
+                    .as_str()
+                    .parse()
+                    .map_err(|e| LangError::WrappingParseIntError(e))?;
+                addr.port = Some(port)
+            }
+            _ => {}
+        }
+    }
+
+    Ok(addr)
+}
+
+#[inline]
+fn parse_host_address(pair: Pair<Rule>) -> RemoteAddr {
+    let s = pair.as_str().trim().to_string();
+    let vs: Vec<&str> = s.split(":").collect();
+    let mut addr = RemoteAddr::default();
+
+    addr.host_name = Some(vs[0].to_owned());
+    if vs.len() == 2 {
+        addr.port = Some(vs[1].parse().unwrap());
+    }
+
+    addr
+}
+
+impl TablePlaceKindContext {
+    fn parse(&mut self, pair: Pair<Rule>) -> LangResult<()> {
+        let r = pair.as_rule();
+        //pre
+        match r {
+            Rule::remote_func => {
+                self.place_kind = TablePlaceKind::Remote(Default::default())
+            }
+            _ => {}
+        }
+
+        for p in pair.clone().into_inner() {
+            self.parse(p)?;
+        }
+
+        //post
+        match r {
+            Rule::ip_address => {
+                self.mut_remote_info()
+                    .map(|place_kind| match parse_ip_address(pair) {
+                        Ok(addr) => Ok(place_kind.addrs.push(addr)),
+                        Err(e) => Err(e),
+                    })
+                    .transpose()?;
```

Comment:
sorry, if we return nothing, why we transpose? ok, we respect the result. then, do we make parse_host_address throw an Error?

Reply:
I think the port can be default, so no error is returned, IP address might get an error, but the host will not get one.

Reply:
The `transpose()?` will return a more specific error when IP address got an error for users.